internal/gengapic/paging.go

```diff
@@ -67,8 +57,33 @@ func (g *generator) iterTypeOf(elemField *descriptor.FieldDescriptorProto) (*ite
 			typ = parent
 		}
 
-		pt.elemTypeName = fmt.Sprintf("*%s.%s", imp.Name, typeName)
-		pt.iterTypeName = typeName + "Iterator"
+		eMsg, ok := eType.(*descriptor.DescriptorProto)
+		if !ok {
+			return nil, errors.E(nil, "cannot find message type %q, malformed descriptor", eType)
+		}
+		if eMsg.GetOptions().GetMapEntry() {
+			var valueField *descriptor.FieldDescriptorProto
+			for _, f := range eMsg.GetField() {
+				if f.GetName() == "value" {
+					valueField = f
+					break
+				}
+			}
+			if valueField == nil {
+				return nil, errors.E(nil, "unusual map entry message: %q", eMsg)
+			}
+
+			if valueField.GetType() == descriptor.FieldDescriptorProto_TYPE_MESSAGE {
+				pt.mapValueTypeName = valueField.GetTypeName()
+			} else {
+				pt.mapValueTypeName = pbinfo.GoTypeForPrim[valueField.GetType()]
+			}
```

Comment:
Same comment here WRT `else` statement contents.

Reply:
This one wound up gaining additional complexity because of message type package disambiguation. The result I think is more clear with explicit if and else instead of default and special case.